chunk_store.go

```diff
@@ -265,39 +265,97 @@ func (c *AWSChunkStore) Put(ctx context.Context, chunks []wire.Chunk) error {
 		return err
 	}
 
-	// TODO: parallelise
+	err = c.putChunks(userID, chunks)
+	if err != nil {
+		return err
+	}
+
+	return c.updateIndex(userID, chunks)
+}
+
+// putChunks writes a collection of chunks to S3 in parallel.
+func (c *AWSChunkStore) putChunks(userID string, chunks []wire.Chunk) error {
+	incomingErrors := make(chan error)
 	for _, chunk := range chunks {
-		err := timeRequest("Put", s3RequestDuration, func() error {
-			var err error
-			_, err = c.s3.PutObject(&s3.PutObjectInput{
-				Body:   bytes.NewReader(chunk.Data),
-				Bucket: aws.String(c.bucketName),
-				Key:    aws.String(chunkName(userID, chunk.ID)),
-			})
-			return err
-		})
+		go func(chunk wire.Chunk) {
+			incomingErrors <- c.putChunk(userID, &chunk)
+		}(chunk)
+	}
+
+	var lastErr error
+	for range chunks {
+		err := <-incomingErrors
 		if err != nil {
-			return err
+			lastErr = err
 		}
+	}
+	return lastErr
```

Comment:
I was curious whether the same pattern appears elsewhere (other than in the two places in this PR)...

https://github.com/weaveworks/frankenstein/blob/20daece52f2ffbe44af01a88a210d84971c7db68/distributor.go#L161
- uses a `for i := 0; i < len(blah); i++` loop - I prefer the `for range blah` trick above; I don't think I've seen it before!
- uses `if err := <-errs; err != nil {` , which is slightly neater

https://github.com/weaveworks/frankenstein/blob/392934e87e30285f9329bde68759daf7d794e530/chunk_store.go#L428
- uses a `for i := 0; i < len(blah); i++` loop
- uses a `select`, which it needs to since it is waiting for result or error